### PR TITLE
Modify color change in text to only sometimes add history

### PR DESCRIPTION
**Pull request by pineapplespatula**
Our history model and modal text states do not currently coexist properly.
to work around this the color set routine which was a nonoptomistic history state
if set in the modal environment should instead append to history.

### CODE REVIEW

File: src/js/actions/type.js [hunk 1]
```diff
@@ -202,7 +202,7 @@ define(function (require, exports) {
      * @param {boolean=} coalesce Whether to coalesce this operation's history state
      * @return {Promise}
      */
-    var updateColor = function (document, layers, color, coalesce) {
+    var updateColor = function (document, layers, color, coalesce, fromSet) {
```
Comment (mcilroyc) [inline src/js/actions/type.js]: nit: this param name is not very meaningful.  Maybe `amendHistory` ?

also: needs to be added to JSDoc

**Comment by pineapplespatula** [comment]
@mcilroyc good points. back to you

File: src/js/events.js [hunk 2]
```diff
@@ -71,6 +71,7 @@ define(function (require, exports, module) {
                     DELETE_LAYERS: "deleteLayersNonOptimistic" // eg: ps deletes the entire layer after last path del
                 },
                 amendment: {
+                    TYPE_COLOR_CHANGED: "typeColorChanged",
```
Comment (mcilroyc) [inline src/js/events.js]: you can't use the same string for both events!  'typeColorChanged'

Comment (mcilroyc) [inline src/js/events.js]: this is causing the bug that ALL setColor calls to use the "amendment" flow

**Comment by mcilroyc** [comment]
Aside from the mistake I mentioned above (which causes a bug in non-modal situations), this modal text flow doesn't really work.  It seems like when you exit the modal text editing state, photoshop "catches up" with several history state events.  This means that we get terribly out of sync.

Also, probably a separate issue: using the text input of the color picker (like, entering "red") does not work very well at all in the modal state.

**Comment by pineapplespatula** [comment]
Photoshop seems to be doing two things oddly. I it is setting two events for each change we make. ( i'm not positivie about this one) and the other is ading a NoOp "Edit Type Layer" history state when we exit the modal state (if a change has been made)  I'm not seeing any issues with the color text input, can you be more specific.

**Comment by mcilroyc** [comment]
Merging this because it prevents a new bogus history state from being captured with every slightest change while in the modal text state (such as arrow key navigation).  

But the problem still remains that making several (>1) color changes while in the modal text state will cause history store to *reset* upon text commit.  This seems to be due to the fact that historyState event notifications from Ps are "batched up" until the modal state is committed.  This causes our history store's internal state to get too far ahead of Ps, and it is not built to withstand the batch of historyState events that it receives some time in the future.

cc @iwehrman as an FYI

**Comment by iwehrman** [comment]
Has somebody filed a watson bug for the problem with history states blocking while the modal tool state is active? This sounds like something that is important to fix for M3. CC @chadrolfs

**Comment by chadrolfs** [comment]
https://watsonexp.corp.adobe.com/#bug=4028040

**Comment by chadrolfs** [comment]
referenced in #1848

**Comment by iwehrman** [comment]
Thx bro :facepunch:

**Pull request merged**